### ISSUE TITLE
Controlling parameters between events via parameter buses

### ISSUE THREAD

**Issue by jarmitage** [musikinformatik/SuperDirt#190]
E.g.:
```haskell
d1 $ pan "0 1" $ s "bev" 
  # panmode "smooth" 1
```

When `# pan "1"` is sent, `# panmode` sets interpolation to `smooth` over `1` cycle duration.

**Comment by telephon** [comment]
good idea.

**Comment by telephon** [comment]
thinking again about this: what would smooth mean here?

**Comment by jarmitage** [comment]
One way this could work is that the panning runs one event or one cycle behind so that it knows where to interpolate to. Thoughts?

And, maybe smooth isn't a useful descriptor and `linear` `ease` etc would be better.

**Comment by telephon** [comment]
ah I see, so you want a pan that moves within each event. It would be possible to add a `panFrom` parameter.

In principle, this would be a possible desire to have for every existing parameter, so I'm not sure if it would be better to introduce some architecture for such a thing.

**Comment by jarmitage** [comment]
> ah I see, so you want a pan that moves within each event. It would be possible to add a panFrom parameter.

Yes, and that's a better suggestion.

> In principle, this would be a possible desire to have for every existing parameter, so I'm not sure if it would be better to introduce some architecture for such a thing.

I completely agree, and I think this would overcome one of the main limitations of the Tidal-SuperDirt relationship cc @yaxu. Should we create a separate issue for that?

**Comment by yaxu** [comment]
We could rename this one. What would we call it. "Parameter envelopes?"

A simple case would be a list of values, equally spaced over the duration of an event (one at start one at end) that supercollider moves between linearly.

We already use `:` as a kind of list separator in the mini-notation (although this needs work to make the elements patternable). 

The interface for that simple case could then be `d1 $ sound "sax" # speed "0.5:1:<0.25 0.5>"`

**Comment by jarmitage** [comment]
I was thinking slightly more generally "parameter interpolation".

It's going to give parameters more independence from the main pattern structure, so there should be a clear way to use the regular approach still.

A notation for this would be great, but maybe we also want a way of doing this with an explicit parameter as well? It's always great having multiple ways of doing things, and sometimes the pattern notation can get quite dense.

What would `# speed "0.5:1:<0.25 0.5>"` do exactly?

**Comment by yaxu** [comment]
Yes I was just trying to start with the simplest case of linear interpolation, 

`# speed "0.5:1:<0.25 0.5>` would make events during the first cycle have a speed starting at 0.5, going up linearly to 1, then down to 0.25. The second cycle it would go down to 0.5 instead. This would probably happen over the duration of the sound.

A helper function could turn 'monophonic' patterns into this kind of envelope e.g. `# pan (slide "0 1")` would be the same as `pan "0:1 1:0"`

We should probably start with finding out how this kind of thing is usually done with supercollider, then looking for a minimal and expressive Tidal UI to support that.

**Comment by yaxu** [comment]
By the way I just checked and dirt used to have a pan_to parameter a couple of rewrites ago, back in 2011 :) http://slab.org/datadirt/datadirt_1.0.0.tar.gz

**Comment by jarmitage** [comment]
I see, although I'm not sure `:` would be very scalable for this use case:
- Writing/reading long patterns with many `:` would be hard
- Trouble combining `:` with `[]`, `,` and `{}` (e.g. `"[0.5:1]:0.25 0.5"`)
- `:` is already used by `grp` to mean separate lists for separate parameters, whereas this proposal suggests separating events within a single parameter pattern

All that being said, I'm afraid I don't have any better ideas off the top of my head.

Until we know a bit more how it's looking on SuperDirt side it might be easier to prototype with additional explicit parameters, and then come back to the UI/notation side later.

And that is a great archeological find - no new ideas under the sun!

**Comment by yaxu** [comment]
My thinking is that both are lists, so should maybe make use of the same syntax, with wider context giving meaning to what the list is of. Additional parameters will likely be needed too, for giving the type of envelope

**Comment by jarmitage** [comment]
Ok, so in this case the "lists" would be lists of patterns that describe interpolation points for the same parameter, with the first list being the "from" pattern, and subsequent lists being "to" patterns?

Wouldn't your example though need to be `"0.5:~ 1:~ ~ <0.25 0.5>"`? I'm struggling to grasp the temporal relations a bit here.

**Comment by yaxu** [comment]
https://github.com/tidalcycles/Tidal/issues/397 related..

**Comment by telephon** [comment]
isn't interpolation just a special case? What you want, in a way, is to be able to change parameters on a running synth, which does the intrpolation internally, e.g. by lagging the parameter.

**Comment by telephon** [comment]
Let's assume we have a synth event called "set", defined like this (we'll need a more general solution):

```supercollider
~dirt.soundLibrary.addSynth(\set, (play: { ~group.set(\freq, ~freq.value) }))
```

Then in tidal, we could make a synth that has a long legato, and while it runs, set it:

```haskell
sound "supergong set set set" # freq "200 700 100 500" # legato 4
```

(I realise that it would be nice to have a legato that is relative to the cycle length)

The synth supergong has no interpolation in its values, but this could work. Still need to test.

Would this lead to a solution in your direction?

**Comment by yaxu** [comment]
That would only work for monophonic synths, unless we made progress on #133.

My suggestion is to send a parameter as an array at trigger time, that sets an envelope on that parameter.

**Comment by yaxu** [comment]
I had a look at the OSCv1.1 spec and couldn't see support for lists

**Comment by telephon** [comment]
> That would only work for monophonic synths, unless we made progress on #133.
> 
> My suggestion is to send a parameter as an array at trigger time, that sets an envelope on that parameter.

The suggestion above is independent of all this. You can set the orbit's group, and this will set the parameters of all synths in that orbit.

So (I thinkt at least) independently of this:
http://opensoundcontrol.org/spec-1_0 down below:

```
[ | Indicates the beginning of an array. The tags following are for data in the Array until a close brace tag is reached.
```

etc.

**Comment by yaxu** [comment]
I see. You might still want different sounds in the same orbit with different frequencies.

I couldn't get the example to work, I get "Instrument isn't found" for "set"

**Comment by telephon** [comment]
I think it should be:
```supercollider
~dirt.soundLibrary.addSynth(\set, (play: { ~synthGroup.set(\freq, ~freq.value) }))
```

**Comment by yaxu** [comment]
Hm I'm still getting the same:
```
-> a DirtSoundLibrary
module 'sound': instrument not found: set
module 'sound': instrument not found: set
module 'sound': instrument not found: set
```
If I run the addSynth line again I get
```
replacing 'set' (1)
-> a DirtSoundLibrary
```
So it seems it's definitely adding it.. Strange.

**Comment by telephon** [comment]
yes, I had the same. I need to investigate why this happens. ~~Also I need to refactor the DirtEvent a little, need to first get the group id so that the synths can grab it.~~

This will be a good first step. If we want to separate the synths, we can think how to do this. There is still the possibility to allocate groups and put synths in groups. Alll this adds complexity though.

**Comment by telephon** [comment]
Here is a first example that actually works:

```supercollider
(
~dirt.soundLibrary.addSynth(\set, 
	(play: { |dirtEvent|
		var group = dirtEvent.orbit.group.asGroup;
		group.set(\freq, ~freq.value) 
	})
);

SynthDef(\supertest, {|out, freq=440, sustain=1, pan |
	var sound = RLPF.ar(Pulse.ar(freq.lag(1)), (freq * 3).lag(2), 0.05);
	Out.ar(out, DirtPan.ar(sound, ~dirt.numChannels, pan))
}).add
);

```

Then you can write:
```haskell
sound "supertest set set set" # freq "200 700 100 500" # legato 4
```

**Comment by telephon** [comment]
What we could do is add control routing busses to superdirt (we already have audio routing busses):

```supercollider
(
~dirt.addModule(\get, { |orbit|
	var bus = orbit.dirt.controlBusses.wrapAt(~fromBus);
	orbit.server.sendMsg("/n_map", ~setParam, bus.index);
}, { ~fromBus.notNil })
)


(
~dirt.addModule(\set, { |orbit|
	var bus = orbit.dirt.controlBusses.wrapAt(~setBus);
	bus.set(~toValue);
}, { ~setBus.notNil })
)
```

And then we could write in tidal, after having defined the right parameter functions:
```
sound "supertest" # setParam "freq" # fromBus "7" # setBus "7" # toValue "403 508 201 703"
```

Both parts, the `setParam` and the `setBus` could happen in completely different orbits.

Currently, we can do this already with audio signal routing, but not with setting controls.

**Comment by telephon** [comment]
P.S. I've added controlBusses now, so the above code should **almost** work.
In tidal:
```haskell
  let fromBus = pI "fromBus"
      setBus = pI "setBus"
      toValue = pF "toValue"
      setParam = pS "setParam"
```

**Comment by telephon** [comment]
OK, here we go:

```haskell
  let fromBus = pI "fromBus"
      setBus = pI "setBus"
      toValue = pF "toValue"
      setParam = pS "setParam"
      fadeTime = pF "fadeTime"
  
d1 $ stack [
      sound "supertest" # sustain 4 # fadeTime 1 # setParam "freq" # fromBus "7" ,
      sound "supersilent*4" # setBus "7" # toValue "403 508 201 703"
      ]
```

SuperDirt:

```supercollider
(
SynthDef(\supertest, {|out, freq=440, sustain=1, pan |
	var sound;
	sound = RLPF.ar(Pulse.ar(freq.lag(1)), (freq * 3).lag(2), 0.05);
	Out.ar(out, DirtPan.ar(sound, ~dirt.numChannels, pan))
}).add;
SynthDef(\supersilent, {|out, freq=440, sustain=1, pan |
	var sound;
	sound = Silent.ar(~dirt.numChannels);
	Out.ar(out, DirtPan.ar(sound, ~dirt.numChannels, pan))
}).add;

~dirt.addModule(\get, { |event|
	var bus = event.orbit.dirt.controlBusses.wrapAt(~fromBus);
	event.orbit.server.sendMsg("/n_map", ~synthGroup, ~setParam, bus.index);
}, { ~fromBus.notNil });

~dirt.addModule(\set, { |event|
	var bus = event.orbit.dirt.controlBusses.wrapAt(~setBus);
	bus.set(~toValue);
}, { ~setBus.notNil })
)
```

Does this go in the right direction?

**Comment by yaxu** [comment]
Instead of sending
```
/play2 "sound" "supersaw" "setParam" "freq" "fromBus" "7"
```
... could tidal instead send the bus value with the name prefixed e.g. by '_'? 
```
/play2 "sound" "supersaw" "_freq" 7
```
Then there could be multiple things set in one message
```
/play2 "sound" "supersaw" "_freq" 7 "_squiz" 8
```
Instead of using '/play2' for setting a bus value, tidal could use a different OSC path, e.g. 
```
/set 7 0.4
```

**Comment by telephon** [comment]
This would work, but would incur a bit of calculation cost on all messages (we need to check every argument name for the prefix, but there is a primitive that is quite efficient).

If this is ok, the nicest way would be:

```
// getting a bus value from a bus named "theFreq"
"_freq" "theFreq"
```

```
// and setting the bus value
"theFreq_" 567
```

But maybe numbers are just fine as well.

Assuming that we do `/set 7 0.4` (which would work much better!).

A very direct way would be to send:
```
/play2 "sound" "supersaw" "freq" "c7"
```

We don't have to change anything on the superdirt side for that really. (Note that this won't work for all parameters, e.g. `n` doesn't work).

**But:** Then we need to guarantee that superdirt has the relevant bus numbers. That probably means that tidal needs to know the busses that superdirt has, and also how many. Superdirt could send them to tidal.

**Comment by telephon** [comment]
The possibility of having synths playing out into audio busses which can then be mapped to arguments is already implemented, but this works differently (there, you mostly don't want to have the same pattern write and read).

**Comment by yaxu** [comment]
Would it be possible to assign a bus number to every possible parameter available, or would that be too many buses?

I'm wondering about a nice interface, maybe this could work:
```
d1 $ sound "supersaw"
  # (bus 7 $ freq "500 600 200)"
```
Then I think that the tidal scheduler could take care of sending the `"freq" "c7"` in the `play2` message, and then separate messages with `/set 7 500`

**Comment by yaxu** [comment]
I have a proof-of-concept for this working now.

This will send one '/play2' message and in theory three `/bus` messages per cycle:
```
d1 $ sound "bd" # (bus 3 $ speed "2 4 3")
```
In practice the `/bus` messages get sent much more often.. I think I'll keep a note of the previous value for each bus and only send it when the value changes.

**Comment by bgold-cosmos** [comment]
Is there a reason to have this in one line (other than convenience) or would there be an equivalent
```
d1 $ sound "bd"

bus 3 $ speed "2 4 3"
```
or perhaps, if buses are dynamically attached to synths
```
d1 $ sound "bd" # bus 3

bus 3 $ speed "2 4 3"

-- now a snare with the same speed variations
d2 $ sound "~ sn" # bus 3 
```
etc...

**Comment by yaxu** [comment]
@bgold-cosmos Interesting, you've made me realise I haven't got my head around this at all!

I'm feeling like there's an underlying issue with the representation which could be holding things back. If you have `sound "a b" # speed "1 2 3"`, and sample it at 20Hz, you can't tell the difference between 'part' onsets in the rightmost pattern and 'part' onsets resulting from the 20Hz query.

Currently the below would return 'parts' every onset of both the left and righthand patterns, _and_ at 20Hz (or whatever the frame size is configured at in tidal):
```
d1 $ sound "bd sn" # (bus 3 $ squiz "2 4 3")
```
.. so there's a lot of extra 'parts' generated. I'm not sure how to get around this.. Maybe query with a very slightly larger (earlier) window and filter out parts that start beforehand?

Separating out into separate patterns could be much more practical, as you could just drive things from the 'whole' onsets in the bus pattern, as usual. Being able to share busses between patterns is a clear win too. But still, being able to specify everything as part of the same pattern feels like it would be nice.. Especially when it comes to things like 'apply a cycle of this bus pattern to every event in this pattern'.

This is possible at the moment:
```
d3 $ bus 5 $ speed "2 4 3"
```
As the above has no 'sound' it doesn't generate `/play2` messages. However this wouldn't generate any events, as the silence would wipe out all the sound events:
```
d2 $ sound "bd sn" # bus 5 $ speed silence
```
There should be some way to add a 'null' event that means that speed still gets set to bus 5, but then the pattern doesn't generate any bus change messages itself..
```
d2 $ sound "bd sn" # bus 5 $ speed null
```
hmm!

**Comment by telephon** [comment]
> Would it be possible to assign a bus number to every possible parameter available, or would that be too many buses?

you'd still want to have separate voices, wouldn't you? This would create a combinatorial explosion. So perhaps it would be ok to just access bus numbers by user defined names?

**Comment by yaxu** [comment]
Yes good point! 

I'm unsure whether to call these "buses", I think people unfamiliar with mixing consoles find that a bit confusing.. But can't think of a better name at the moment.. maybe "dial"? As 'dial' implies a (rotary) control for something, but also communication - 'dialling something up' (dead metaphor from old fashioned telephones)?

Also currently the main branch identifies buses by number rather than by name. I'm not sure whether I'd prefer names or not.. 

Numbers woudl be like this:
```
d2 $ sound "sax" # (dial 5 $ squiz "3 5 2")
```
The name would probably often be the name of the parameter that's being controlled, e.g.:
```
d2 $ sound "sax" # (dial "squiz" $ squiz "3 5 2")
```
Except when you wanted to control two squizzes separately..
```
d1 $ sound "bev" # (dial "squiz1" $ squiz "3 5 2")
d2 $ sound "sax" # (dial "squiz2" $ squiz "0 3 1")
```
Hmm.. Maybe there could be a primed version of every parameter that constructed a unique bus name from the pattern id and the parameter name that was being 'dialled'
```
d3 $ sound "bev" # (squiz' "3 5 2") -- would send to bus named "squiz3"
```

**Comment by yaxu** [comment]
> Assuming that we do `/set 7 0.4` (which would work much better!).
> 
> A very direct way would be to send:
> 
> ```
> /play2 "sound" "supersaw" "freq" "c7"
> ```
> 
> We don't have to change anything on the superdirt side for that really. (Note that this won't work for all parameters, e.g. `n` doesn't work).

When you say we don't really need to change anything for this, do you mean it's easy to implement or already implemented? This would be really great to play with!

**Comment by telephon** [comment]
The pieces are there, they need to be assembled in tidal.

You can try this.

Set up something that sends to superdirt messages of the following form:

```
"/setControlBus" index value 
```
// where index is an integer and value is some float

then let's say you have the following synth:

```supercollider
(
SynthDef(\xyz, { |out, sustain = 1, ttt|

	var sound, env;
	env = Env.perc(releaseTime: sustain - 0.01).kr;
	sound = SinOsc.ar(400 * (ttt + 1), SinOsc.ar(13 * (ttt + 1)));
	Out.ar(out, sound * env);
	
}).add

)
```

And from tidal you should be able to run:
```
d2 $ sound "xyz" # ttt "c_0"

```

Mapping the control "ttt" to the bus zero (this assumes, for now that the busses match their indices in superdirt, but we can make sure that this is the case later).

Then, you have continuous control from tidal, if you set up something that sends a message like

```
"/setControlBus" 0 <some random value between 0 and 1> 
```


My test code in supercollider was:

```supercollider


SuperDirt.default = ~dirt;

(
SynthDef(\x, { |out, sustain = 1, ttt|

	var sound, env;
	env = Env.perc(releaseTime: sustain - 0.01).kr;
	sound = SinOsc.ar(400 * (ttt + 1), SinOsc.ar(13 * (ttt + 1)));
	Out.ar(out, sound * env);
	
}).add

)

(type: \dirt, sound: "x", ttt: 7).play
(type: \dirt, sound: "x", ttt: "c_0").play

~dirt.controlBusses.first.bus

(
fork { 
	loop {
		0.1.wait;
		~dirt.setControlBus(0, 5.0.rand.postln);
	}
}
)
```

**Comment by yaxu** [comment]
Exccellent thanks very much! This is now working with the tidal main branch:
```
d1 $ sound "xyz" # (dial 0 $ ttt (segment 16 perlin))
```
This isn't yet:
```
d1 $ sound "bev" # (dial 0 $ squiz (segment 16 perlin)) # legato 1
```

**Comment by telephon** [comment]
nice.
What should the second line do? What is missing?

**Comment by yaxu** [comment]
I thought it should modulate the 'squiz' effect in the same way as the 'ttt' synth parameter.

This works:
```
d1 $ note "c" # sound "supermandolin" # legato 4 # (dial 0 $ detune "[0 1 4 2]*3")
```
but not if I pick a different bus:
```
d1 $ note "c" # sound "supermandolin" # legato 4 # (dial 0 $ detune "[0 1 4 2]*3")
```
This seems to be working:
```
d1 $ sound "clap:4" # delay 0.5 # (dial 0 $ delayt "0.3333 0.25") # delayfb 0.97 # lock 1
```
.. although if I set a low value for delaytime, like 0.02, it works once then goes a bit funny.

This is working too:
```
d1 $ sound "sax" # (dial 0 $ djf "0.86 0.25 0.125")
```

Fun!

**Comment by telephon** [comment]
Great! Before we start testing which parameters work an which don't (e.g. `n` won't work for sure), there is one issue that needs to be dealt with.

In principle, the actual bus numbers are not in the range `0..numControlBusses-1` unless you really start superdirt with a freshly booted server. So the argument to `dial` is really an index into the array of control busses, and not directly the bus ID.

The most efficient way to fix this is a handshake, where tidal asks for the numbers and superdirt sends them to tidal. Then tidal can also set the busses directly on the server, without going through sclang.

So I could define an OSC command "/handshake" that sends back a json file with some information like the bus indices and server address. This could be extended in the future.

**Comment by telephon** [comment]
> .. although if I set a low value for delaytime, like 0.02, it works once then goes a bit funny.

Yes, `SwitchDelay` can't be modulated. We need to protect it from such a mess …

**Comment by yaxu** [comment]
This sounds good. A failing handshake with a useful error message could also help people work out what is wrong with their system. A superdirt or protocol version number would be useful too. What would happen if someone had an older version of superdirt that didn't understand the handshake command? Would it be silent or would it send back an error message?

**Comment by telephon** [comment]
That older version would be silent. So you'd need to wait for a handshake reply and then post that either the version is too old, or the network connection has failed.

**Comment by yaxu** [comment]
Ok thanks. Why json in particular, as opposed to an OSC message with name-value pairs? json would be an additional dependency on the tidal side.

**Comment by telephon** [comment]
I see, yes, I am fine with key value pairs.

**Comment by yaxu** [comment]
We've talked about storing audio file metadata as json though. Perhaps superdirt could still send that info as name-value OSC pairs, unless that needs more structure.

**Comment by telephon** [comment]
Ok, almost done. So tidal would ask for `"/handshake"` and superdirt replies `"/handshake_reply"`, and generates this data (for now):

```
handshakeReplyData {
		^[
			\controlBusIndices, controlBusses.collect { |x| x.index },
			\serverHostname, server.addr.hostname,
			\serverPort, server.addr.port
		]
	}
```

from the list of indices you can then do two things:

1. you can send `"/c_set", index, value`control messages directly to the server (where value is a 32 bit float). Just make sure the timing is correct.
2. you can embed `c<bus index>`) as synth parameter, e.g. `"ttt" c34`.

**Comment by telephon** [comment]
ah, one thing. Do you have array values implemented on the tidal side? 
```
"controlBusIndices", "[", 7, 8, 9, "]"
```

**Comment by telephon** [comment]
http://opensoundcontrol.org/spec-1_0
(see: additional argument types )

**Comment by yaxu** [comment]
Sadly not, it's commented out here: https://hackage.haskell.org/package/hosc-0.18.1/docs/src/Sound.OSC.Datum.html#osc_types_optional

Perhaps you could just send multiple integers, in this case `controlBusIndices 7 8 9` with typetags `siii`. In this case it's unambiguous, if names are always strings.

**Comment by telephon** [comment]
then we would have to add extra (nonstandard) information about the length of every value:

```
"controlBusIndices" 3 7 8 9 "serverHostname" 1 "168.2.2.2" "serverPort" 1 57110
```

(update)

> if names are always strings.

I see, yes, probably. Unless we want to send arrays of strings as values. Can we exclude that case?

We could send the names with a specific character, like `"#controlBusIndices` or similar. This could make it relatively safe for the future.

@yaxu just let me know which would be the best choice …

**Comment by jarmitage** [comment]
This is shaping up to be great!

Going back to the original issue title/topic, re interpolation:

> isn't interpolation just a special case? What you want, in a way, is to be able to change parameters on a running synth, which does the intrpolation internally, e.g. by lagging the parameter.

I assume at this point that the interpolation aspect would just be internal to whatever synth/effect is being controlled? It would still be cool to somehow do this for a smooth panning effect:

```haskell
d1 $ sound "bev" # (dial 0 $ pan "0 1" # interpolate "ease")
```

Should this be a separate issue now?

**Comment by telephon** [comment]
>I assume at this point that the interpolation aspect would just be internal to whatever synth/effect is being controlled?

Yes, for now.

> It would still be cool to somehow do this for a smooth panning effect:

I think this is for later. This would require a synth that writes on the control bus.

**Comment by jarmitage** [comment]
Trying this out, it seems pretty unstable for me. I get frequent dropouts if I do:

```haskell
d1
  $ loopAt 8 $ sound "bev"
  # (dial 0 $ segment 16 $ djf (slow 4 $ range 0.1 0.9 sine) # pan (range 0.1 0.9 sine))
  # legato 1
```

Chaining params to a `dial` doesn't work? 

Exciting though!

Maybe this is just something to do with `djf`? Or maybe there needs to be at least a tiny bit of parameter smoothing/interpolation?

```haskell
p1
  $ loopAt 8 $ sound "bev"
  # (dial 0 $ djf "[0.1 0.9]*4")
  # legato 1
```

Can't get `pan` and `gain` to work. `pan` is weird: 
1. It only listens to the first event
2. It scales the parameter between -1 and 1

```haskell
p1
  $ sound "bev"
  # (dial 0 $ pan "-1 0.25 0.5 0.75 1")
  # legato 1
```

Testing a few more...

Work:
- `room`, `size`, `leslie`, `distort`

Doesn't:
- `shape`, `crush`, `tremdp`, `tremr`, `coarse`, `vowel`
- `waveloss`, `squiz`, `freeze`

Only listens to 1st event:
- `gain`, `pan`, `lpf`,

**Comment by telephon** [comment]
>Maybe this is just something to do with djf? Or maybe there needs to be at least a tiny bit of parameter smoothing/interpolation?

Yes, sure, all the `SynthDef`s that have filters can blow up, and others will sound bad. The easiest solution is to smooth the values, but we may have to think about some uniform way to do that.

> Can't get `pan` and `gain` to work. `pan` is weird:
> 
>     1. It only listens to the first event

Yes, for efficiency reasons, so far, all those parameters are locked.

> 
>     2. It scales the parameter between -1 and 1

This is what it should do. The pan range in the `Pan2` UGen (and all the other pan ugens) is -1..1

**Comment by telephon** [comment]
I've just checked with the vowel and coarse SynthDefs – making all parameters modulatable would cost around 10 % cpu, about equally in the cases of stereo and 20 channels. 

This increase in load affects the whole of the synth, so we need to think about whether it is worth it. Theoretically we could make separate versions and tidal could send a flag, but this sounds like a maintenenance hell to me.

**Comment by telephon** [comment]
Another issue to check would be that one may want to have different lags with different parameters. This would be tricky in terms of naming …

**Comment by jarmitage** [comment]
Ok. FWIW I've wanted this feature for years primarily for using with `pan` and `gain`. So if there was a choice to be made CPU-wise, for me I'd rather have this feature for those two params and nothing else, rather than for a couple of random effects.

> This is what it should do. The pan range in the Pan2 UGen (and all the other pan ugens) is -1..1

`pan` in Tidal is 0-1 normally (`d1 $ pan "0 1" # s "bd"` will be left then right) so not sure why it's being scaled in this case.

**Comment by BenjaminWynn** [comment]
Just jumping in to say that I've wanted this for numerous effects, the ones I remember most recently are the distortions, filters and gain. I'd vote for the cpu hit to have it work for all FX personally, if that makes sense. It might be confusing if it works for only some imho. 

Also, really looking forward to this!

**Comment by BenjaminWynn** [comment]
PS: will this also work for my own SC SynthDefs? Like say I have a wavetable synth, and I want one long held note, but different wavetable positions sent to SC without retriggering during that note.

**Comment by yaxu** [comment]
I think 10% cpu is definitely worth it (depending on what that's 10% of). It'd be a deep and welcome change in the relationship between tidal and sound. If there was a startup option to switch it on and off for lower powered systems that'd be perfect.. Or even an easy way to manually switch off unused effects to reduce CPU. But there's always classic dirt for those lower powered systems

**Comment by telephon** [comment]
> If there was a startup option to switch it on and off for lower powered systems that'd be perfect.. Or even an easy way to manually switch off unused effects to reduce CPU. 

Possible, but makes the code harder to read. I'd rather not do that.

One thing I'm wondering though. I can easily make various parameters modulatable and add a lag to them, and make this lag accessible. But each synth will have one lag for all its parameters. What if you want to modulate one faster than another? Also, calculating the lag is part of the cpu-load.

An alternative would be not to lag the synth controls at all and instead make an interface that allows tidal to send commands that crossfade the value of the control bus instead. Then, if you want to fade at different speeds, you just don't map to the same bus. This makes more sense to me.

Actually, there is already a way to do it, as it turns out.

Try send OSC messages to scsynth (the server) like the following:
```
"/s_new", "system_setbus_control_1", -1, 0, 1, "out", <busNum>, "values", <value>, "fadeTime", <fadeTime>, "curve", 1
```

For example, crossfade control bus 3 in 0.2 sec to the value 100:
```
"/s_new", "system_setbus_control_1", -1, 0, 1, "out", 3, "values", 100, "fadeTime", 0.2, "curve", 1
```

Can you try that?

**Comment by telephon** [comment]
> PS: will this also work for my own SC SynthDefs? Like say I have a wavetable synth, and I want one long held note, but different wavetable positions sent to SC without retriggering during that note.

If your parameters are control rate (that is, if you haven't added them to the list of `[\ir, \ir]` specs at the end of the `SynthDef`, then yes, they'll work. You just need to take care not to blow up filters.

**Comment by telephon** [comment]
> Ok. FWIW I've wanted this feature for years primarily for using with `pan` and `gain`. So if there was a choice to be made CPU-wise, for me I'd rather have this feature for those two params and nothing else, rather than for a couple of random effects.

Actually, I realise now that it is the other way round. Because `pan` is standard `-1 … 1` in supercollider, sclang converts them for you. But we've kept the synths still in this bipolar range in order to make them consistent with the supercollider signal processing standards. But because the direct modulation is done by bus mappings, you would have to use that range from tidal. Or tidal would have to check and convert that range, which would be better anyhow.

The situation for `gain` is more complicated even. `gain` translates to `amp`(which you can also directly specify), using the following formulation:

```supercollider
~amp = pow(~gain.value.min(2) + ~overgain.value, 4) * ~amp.value;
```

We do not have this calculation on the busses. So you'd need to map the `amp` value. In Supercollider, multiplication on a symbol is a no-op, so that `~amp = 0.37577 * "/c2"` wil evaluate as `"/c2"`, which would map the synth to control bus with index 3.


> 
> > This is what it should do. The pan range in the Pan2 UGen (and all the other pan ugens) is -1..1
> 
> `pan` in Tidal is 0-1 normally (`d1 $ pan "0 1" # s "bd"` will be left then right) so not sure why it's being scaled in this case.

**Comment by yaxu** [comment]
So instead of sending `/setControlBus <busNum> <value>` to sclang, I send the `/s_new` to scsynth? That should be straightforward, thanks! I'm in Munich at the moment but will try to find some time in the evening.

**Comment by telephon** [comment]
Ah I wish I could be there.

**Comment by telephon** [comment]
Here is a branch that opens up effects for modulation and we can have a handshake.

https://github.com/musikinformatik/SuperDirt/pull/210

**Comment by yaxu** [comment]
The handshake is working with this tweak: https://github.com/musikinformatik/SuperDirt/pull/211

I get all the numbers 0 .. 127, having started with a cleanly booted server. I don't know how these busses work - is it possible that any of them become unavailable while superdirt is running?

**Comment by telephon** [comment]
>  I don't know how these busses work - is it possible that any of them become unavailable while superdirt is running?

You can use this list as a namespace for bus indices. They will not become unavailable while superdirt is running, but when superdirt is restarted. So you need to repeat the handshake in some intervals, I suppose.

**Comment by yaxu** [comment]
A thought - with trigger messages going to via sclang, and control messages going straight to scsynth, there's the difficulty of arranging for them to arrive at the right time in relation to one another. In an ideal world, I guess they would happen at exactly the same moment in time, and then processed in the right order (update bus, trigger sound).

Does scsynth honour buffer timestamps? If I look with a packet sniffer, I see the /play2 going to sclang, then an _empty_ bundle going from sclang to scsynth.. At least that what it seems in wireshark, a bundle without any messages in..

**Comment by telephon** [comment]
>Does scsynth honour buffer timestamps? 

Yes, of course. The timestamps should actually be almost the same, because we pass the latency parameter on to the message forwarding. I am not sure if the calculation is correct though.

> In an ideal world, I guess they would happen at exactly the same moment in time, and then processed in the right order (update bus, trigger sound).

Well, there is a problem in OSC that you can't order timestamped bundles other than through timestamps. But in our case, the order doesn't matter, as long as the timing is correct. The bus can be updated in parallel.

>If I look with a packet sniffer, I see the /play2 going to sclang, then an empty bundle going from sclang to scsynth.. 

Hm, that is strange. We are not sending an empty bundle. Do you see any other bundle from sclang to scsynth? Sure that it is empty? Maybe wireshark doesn't display the bundle contents?

**Comment by telephon** [comment]
> > Does scsynth honour buffer timestamps?
> 
> Yes, of course. The timestamps should actually be almost the same, because we pass the latency parameter on to the message forwarding. I am not sure if the calculation is correct though.

I've just tested that part, it is not 100% the same, but correct up to a tenth of a millisecond, so we can rely on that bundles  from tidal being sent to sclang and to scsynth with the same timestamp should arrive at the same time (up to the precision of around 10^-4 sec). 

Edit: new synths are instantiated at control bus boundaries, which are by default a millisecond apart, so this is still fine. I do wonder where the difference comes from …


You can test it on your system [EDITED]:
```supercollider
(
n = NetAddr("127.0.0.1", NetAddr.langPort);

OSCFunc({ |msg, time|
	n.sendBundle(time - Main.elapsedTime, ["/indirect"]);
}, "/pass_on");
OSCFunc({ |msg, time|
	"indirect time: %".format(time.cs).postln;
}, "/indirect");
OSCFunc({ |msg, time|
	"direct time: %".format(time.cs).postln;
}, "/direct");
)

(
n.sendBundle(0.5, ["/direct"]);
n.sendBundle(0.5, ["/pass_on"]);
)

```

**Comment by yaxu** [comment]
Here's where I'm working on the tidal side: https://github.com/tidalcycles/Tidal/compare/topic-handshake

So far it retrieves and stores the list of available busses, although isn't actually using them yet.

**Comment by telephon** [comment]
Great! The idea of the `&` as a prefix is that in the future this would still allow us to send arrays of strings.

**Comment by telephon** [comment]
One thing we may want to consider is that superdirt allows you to delay a bundle according to some parameters, especially `# lag`, `# offset`, together with `# speed`:

```supercollider
~latency = ~latency + ~lag.value + (~offset.value * ~speed.value.abs);
```

I am not sure how that factors through the `dial` function. It may be desirable to have such a lag relative to the bus values. But in other cases, one may want to lag dialed synths.

**Comment by yaxu** [comment]
> Great! The idea of the `&` as a prefix is that in the future this would still allow us to send arrays of strings.

Super. I guess if we really want to send a string that starts with an `&` we just have two `&`s. So `"&foo&bar"` would be `"&&foo&bar"` (no need to escape the one that isn't at the start)

**Comment by telephon** [comment]
OK, done!

**Comment by telephon** [comment]
Ups, misread. I'll undo.

**Comment by yaxu** [comment]
This is a little strange. As I mentioned elsewhere (I think on the PR), everything is working, but wireshark is having trouble rendering the contents of the osc bundle that superdirt sends from sclang to scsynth:

![image](https://user-images.githubusercontent.com/27940/95217061-12df3f00-07f3-11eb-938b-d63ed6529bde.png)

Looking at the above, the bundle seems to set the size of the message to only 16 bytes.

For comparison he's the bundled osc sent from tidal:

![image](https://user-images.githubusercontent.com/27940/95217413-710c2200-07f3-11eb-9532-debf48d68b1a.png)

This is in supercollider "3.11.1-rc1".

**Comment by yaxu** [comment]
Well I don't really understand what's going on, but probably it's just wireshark failing to understand a bundle with more than one element inside.

**Comment by yaxu** [comment]
Where superdirt receives a message from an address it hasn't heard from before, would it be easy for it to detect that, and send a friendly `/hello` back? Then tidal would know that it is talking to a new instance of superdirt (probably because it has been restarted) and that it should shake its hand again (like the old days).

**Comment by telephon** [comment]
>  but wireshark is having trouble rendering the contents of the osc bundle that superdirt sends from sclang to scsynth

Probably sclang is sending the bundle as a blob. ([open sound control](https://web.archive.org/web/20200217222613/http://opensoundcontrol.org/spec-1_0) type tags them as "b").

**Comment by telephon** [comment]
> Where superdirt receives a message from an address it hasn't heard from before, would it be easy for it to detect that, and send a friendly `/hello` back? Then tidal would know that it is talking to a new instance of superdirt (probably because it has been restarted) and that it should shake its hand again (like the old days).

Done. I've changed the command names to avoid possible collisions:
```
/hello_tidal
/hello_tidal_reply
/handshake_tidal
/handshake_tidal_reply
```

**Comment by telephon** [comment]
Ah maybe they should be backwards. 

```
/tidal/hello
/tidal/hello/reply
/tidal/handshake
/tidal/handshake/reply
```

is that better?

**Comment by yaxu** [comment]
I wonder if `/superdirt`, or maybe the more economical `/dirt` is better. It seems to me that the protocol is superdirt, rather than tidal specific. For example conductive might want to make use of this.

**Comment by yaxu** [comment]
Then it could be
```
> /dirt/play "s" "bd"
< /dirt/hello
> /dirt/hello/reply tidal
> /dirt/handshake
< /dirt/handshake reply <data burst>
```

**Comment by telephon** [comment]
Ok, that makes sense.

**Comment by yaxu** [comment]
That replaces `/play2` with `/dirt/play` although it would be good to keep supporting `/play2` for backward compatibility

**Comment by telephon** [comment]
yes, I'm keeping that, too.

**Comment by yaxu** [comment]
Right this is working now https://github.com/tidalcycles/Tidal/compare/topic-handshake

If I run `Bus.control(s, numChannels: 1)` a few times before starting superdirt, the bus numbers get offset fine

It's a bit messy sending different things to different ports, I think I've built up some technical debt with hacks..

I'm not happy with the interface on the tidal side. I think `pan (dial 0 "0 1")` would be possible, and better than the current `(dial 0 $ pan "0 1")`. `pan (dial "0 1")` might also be possible, keeping track of bus numbers on the pattern but would also be a bit messy..

This just works with numbers I guess? Maybe it's possible to have a version of `vowel` that works with numbers rather than letters, maybe even with a float so it can tween between vowels?

**Comment by jarmitage** [comment]
`pan (dial 0 "0 1")` looks good.

Although, I assume that means it's not possible to send multiple effects to one dial?

What about shortcuts for `dial0` etc?

**Comment by yaxu** [comment]
I think having multiple effects read from the same bus depends on this issue https://github.com/tidalcycles/Tidal/issues/721

There's quite a bit of plumbing involved basically..

**Comment by yaxu** [comment]
I'm trying to think of an UI for this. Maybe separating out `send` and `recv` is best
```
d1 $ sound "longsound" # pan (recv 0)
d2 $ sound "anotherlongsound" # pan (recv 0)
d3 $ send "0 1"
```
Then instead of `dial` we could make an alias for every buss-able parameter, e.g:
```
d1 $ sound "longsound" # panB 0 "0 1"
```
Ideally tidal could take care of selecting an unused bus for you, so you could just do:
```
d1 $ sound "longsound" # pan_ "0 1"
```
It could be that we prefer this to be the default behaviour.. In which case `pan` would be the one that gives separable controls and `pan_` could maintain the onset-driven behaviour..

**Comment by telephon** [comment]
Reading all this, what about having a bus number allocator in tidal that you can use to declare busses, and name them? Then refering to the name (which could be numbers as well, and the numbers `0..9` could be reserved for `dial`) would give you that bus. Also, a variant of `dial` without a bus name would implicitly allocate and deallocate a bus number.

**Comment by telephon** [comment]
Some inherent problems to be solved are, off my head:

- how to map the same modulation shape to different parameter ranges
- how to switch between mapping and direct sending of values within a pattern
- how to deallocate bus numbers once an implicit dial falls out of use

**Comment by yaxu** [comment]
Hm, so instead of giving the user the cognitive load of working out what number is for what, we could name them rather than numbering them?

There's already a way to name patterns globally, e.g.
```
setF_ "hello" "0 1 2"
d1 $ n "^hello" # s "drum"
```
adding a separate layer of named patterns would get confusing I think. But I guess we could use the same dictionary for storing the patterns, whether or not they are sent to busses. Hmmm

To expand on your implementation problems, questions for the UI..
- how to specify either onset-driven or continuous updates of event parameters (I think this is restating your second problem)
- how to share busses across pattern parameters (or do we wastefully give each instance of use its own bus, so that we can e.g. easily solve problem 1 - map different instances to different ranges?)
- how to specify an 'anonymous bus' that is just used in one place and so doesn't need naming

**Comment by yaxu** [comment]
It would be generally nice to be able to name a part of a pattern for use elsewhere. e.g.
```
d1 $ sound "bd sn" # pan (label "fred" "0 1 2")
d2 $ sound "cp" # speed "^fred"
```
With this way of naming things, then we can share patterns between anonymous busses
```
d1 $ sound "bd sn" # panbus (label "fred" "0 1 2")
d2 $ sound "cp" # speedbus "^fred"
```
That sends the same values to two different busses, wasteful but flexible, as we can quickly change the mapping so they aren't the same:
```
d2 $ sound "cp" # speedbus ("^fred" |+ 1)
```
and if we really want to specify the bus for efficiency reasons, we could do that by number:
```
d1 $ sound "bd sn" # panbus (send 3 "0 1 2")
d2 $ sound "cp" # speedbus (recv 3)
```
to get efficient, you have to understand the abstraction, so it's a tradeoff..

The simplest case is where you don't want to share a parameter:
```
d2 $ sound "cp" # speedbus "2 3 4"
```

(well speed can't actually be modulated like this but hey)

**Comment by telephon** [comment]
>That sends the same values to two different busses, wasteful but flexible, as we can quickly change the mapping so they aren't the same:

Ah yes, this fixes the first issue. It's not that wasteful, they are just control busses.

Btw. by default, scsynth allocates 16384 control busses. No problem to take a chunk. That said, if we take more, perhaps we should just get a contiguous range, so that the handshake doesn't have so much redundancy.

> The simplest case is where you don't want to share a parameter:

Looks good. How would you specify the other parameters for the bus signal, like `fadeTime` and `curve`? Especially if you have two different parameters like `freqbus` and `panbus` in one pattern, and you want each of them have different fade times?

Btw. we won't be able to use `speed` on a bus easily. There is too much calculation that happens on the discrete value in sclang.

> ```
> d1 $ sound "bd sn" # panbus (label "fred" "0 1 2")
> d2 $ sound "cp" # speedbus "^fred"
> ```

would the following work, too?

```
d1 $ sound "bd sn" # pan (label "fred" "0 1 2")
d2 $ sound "cp" # pan "^fred 2 0 ^fred 3"
```

**Comment by jarmitage** [comment]
> But I guess we could use the same dictionary for storing the patterns, whether or not they are sent to busses. Hmmm

I noticed that convergence. I think it's worth trying if possible! This could have awesome consequences...

**Comment by yaxu** [comment]
> Ah yes, this fixes the first issue. It's not that wasteful, they are just control busses.

Good!

> Btw. by default, scsynth allocates 16384 control busses. No problem to take a chunk. That said, if we take more, perhaps we should just get a contiguous range, so that the handshake doesn't have so much redundancy.

Ok. There's also the problem of what happens if more than one tidal (or other system) wants some control busses. I guess tidal should identify itself with a name when it handshakes, and superdirt sends back a different range for senders with different names?

> Looks good. How would you specify the other parameters for the bus signal, like `fadeTime` and `curve`? Especially if you have two different parameters like `freqbus` and `panbus` in one pattern, and you want each of them have different fade times?

Hm, I'm not sure. What are the possible values for `curve`?

> Btw. we won't be able to use `speed` on a bus easily. There is too much calculation that happens on the discrete value in sclang.

Ok. How about `note`?

> would the following work, too?
> ```
> d1 $ sound "bd sn" # pan (label "fred" "0 1 2")
> d2 $ sound "cp" # pan "^fred 2 0 ^fred 3"
> ```

Yes I think so.

Actually implementing this in Tidal is a bit of a headache.
```
d5 $ sound "sax" # squizbus "2 3 4"
```
It's hard to see how `squizbus` could request a unique bus number without things getting very fiddly (I've been looking into putting metadata on the pattern, event and values).. But maybe such a bus could simply form a unique identifier of its control name prefixed by the id of the pattern, in this case `5squiz`. Then that gets allocated to a bus number by the scheduler, which picks one that hasn't been used yet, or failing that, steals the one that has been used least recently. I wonder what an efficient datastructure for that bookkeeping would be..

**Comment by telephon** [comment]
> I wonder what an efficient datastructure for that bookkeeping would be..

maybe something like this or below: ~~https://github.com/musikinformatik/SuperDirt/issues/190#issuecomment-706570707~~

https://github.com/supercollider/supercollider/blob/develop/SCClassLibrary/Common/Control/Engine.sc#L94

**Comment by telephon** [comment]
>Ok. There's also the problem of what happens if more than one tidal (or other system) wants some control busses. 

I see, yes so far we had no real problem with tandem usage, because each event was atomic. I can allocate the busses dynamically on request. What should be the reply to tidal if that handshake has failed?

**Comment by telephon** [comment]
>Ok. How about `note`?

No note.

These do not work:
```
midinote
note
n
octave
begin 
end
sustain
legato
loop
unit
length
fadeTime
fadeInTime
speed
endSpeed
gain
overgain
channel
lag
offset
pan (wrong mapping in tidal)
```

You can use `freq`, I think.

In order to make (most of) these work, we'd need to move the stuff that is done in https://github.com/musikinformatik/SuperDirt/blob/develop/classes/DirtEvent.sc and the defaults in https://github.com/musikinformatik/SuperDirt/blob/7ae484bed7d3747a2b9ce3d0e07ee17e8e41d4a2/classes/DirtOrbit.sc#L176 to tidal, and separate them in superdirt if you send events from sclang.

Quite a bit of work, but feasible.

**Comment by telephon** [comment]
>It's hard to see how squizbus could request a unique bus number without things getting very fiddly 

What if every pattern carries with it a unique hash?

**Comment by yaxu** [comment]
> maybe something like this or below: [#190 (comment)](https://github.com/musikinformatik/SuperDirt/issues/190#issuecomment-706570707)

Did you mean a different link?

**Comment by telephon** [comment]
https://github.com/supercollider/supercollider/blob/develop/SCClassLibrary/Common/Control/Engine.sc#L94

**Comment by telephon** [comment]
If you wanted, you could also ask for each bus separately from dirt via OSC. Then there is no need for writing an allocator.

**Comment by yaxu** [comment]
Thanks! I guess a network ping pong each time a bus is used would slow things down though?
I am super nervous about using strings as keys though. I've realised I have an irrational fear of string comparison. I guess it's not a super efficient thing to do, and haskell's default string representation is known to be a bit slow. Maybe it's worth switching Tidal to a more efficient string type https://mmhaskell.com/blog/2017/5/15/untangling-haskells-strings
That blog post suggests that it's only string manipulation that's inefficient with basic strings, and I don't think we do much of that..

**Comment by telephon** [comment]
> I guess a network ping pong each time a bus is used would slow things down though?

You'd need it only the first time you define a pattern that uses the bus. But maybe yes.

>I am super nervous about using strings as keys though. 

I also think that strings are not necessary. Numbers are fine, if there is an implicit number allocator.

**Comment by telephon** [comment]
> I am super nervous about using strings as keys though. I've realised I have an irrational fear of string comparison. I guess it's not a super efficient thing to do, and haskell's default string representation is known to be a bit slow.

Not sure if we need it at all, but you could use `Symbol`, couldn't you? Usually, symbols have a unique identity, so they are very fast in checking for that.

**Comment by yaxu** [comment]
I got a bit bogged down in the technicalities of this, so keen to get a simple version released..

@telephon by `Symbol`, do you mean this:
  https://hackage.haskell.org/package/symbol

Or is this something on the SuperCollider side?

Anyway to reiterate, a simple start would be to support something like this to allocate a bus (denoted by the suffix `_`):
```
d1 $ sound "sax" # vowel_ "a e i o"
```
That reaches the scheduler as the identifier `d1vowel` or similar, which will retrieve or allocate a bus number, and will send it but only if it is different from the previous value that was sent.

I think that's enough for now. It might be nice to be able to pattern the bus ids, so that one parameter could switch between receiving values from different busses, but that mgiht be taking it too far..

**Comment by telephon** [comment]
> @telephon by `Symbol`, do you mean this:
> https://hackage.haskell.org/package/symbol
> 
> Or is this something on the SuperCollider side?

I think it is better if tidal can maintain the binding between names and bus indices, unless that's really too much work. Otherwise, superdirt can also do this, as long as tidal makes sure the timing is correct and names are marked as freed after use somehow.

>That reaches the scheduler as the identifier d1vowel or similar, which will retrieve or allocate a bus number, and will send it but only if it is different from the previous value that was sent.

So this is now for writing the bus signal, right? Then yes, since the synth lives across the messages it will keep the mapping. But this is just a matter of optimisation.

For the synth that reads the bus signal, you'll have to send a `"c12"` like parameter in each message.

I hope I understood you correctly.

**Comment by yaxu** [comment]
Yes just a matter of optimisation, but currently without this tidal will generate a lot of spurious identical bus messages at 20hz. Mainly I was just reiterating things after returning to this after a break.

Though this turns out to be difficult:
```
d1 $ sound "sax" # vowel_ "a e i o"
```
Because of the book keeping needed to support multiple versions of the same parameter e.g.:
```
d1 $ stack [sound "sax" # vowel_ "a e i o", sound "kick" # vowel "a o i", sound "clap" # vowel_ "e a u"]
```
I think they have to have unique identifiers, so I might go back to using 'dial'. I find this fiddly though..
```
d1 $ stack [sound "sax" # (dial 3 $ vowel "a e i o"), sound "kick" # vowel "a o i", sound "clap" # (dial 4 $ vowel "e a u")]
```
So perhaps something like this is better
```
d1 $ stack [sound "sax" # vowelbus 3 "a e i o", sound "kick" # vowel "a o i", sound "clap" # vowelbus 4 "e a u"]
```
That doubles the size of Params.hs.. Or I think we could use template haskell to generate the declarations and make it much smaller than it is.

**Comment by telephon** [comment]
The explicit version (dial) is a good intermediate step, maybe. To go further, it seems that you need a way to identify the start and end of a particular subpattern?

Then the ids for the dials could be hidden under a hood. And they could be overridden as well to make cross pattern communication possible.

**Comment by yaxu** [comment]
Ok I'm playing around with this for the first time, e.g.:
```
d1 $ sound "sax"
  # resonance 0.2
  # cutoffbus 3 (fast 4 $ segment 32 $ range 0 3000 saw)
  # ampbus 5 "[1.5 0]*8"
```
This sounds nice but there's seems to be a strange lag. If I comment out the ampbus line for example, it takes a couple of seconds for it to fade to 1, and then if I uncomment it, it similar amount of time to fully take effect..

**Comment by telephon** [comment]
Maybe let's reconsider what happens on the OSC level? What are the messages that are sent with `# ampbus 5 "[1.5 0]*8"`?

**Comment by yaxu** [comment]
Ok this simpler pattern has the same kind of behaviour:
```
once $ sound "sax"
  # resonance 0.2
  # cutoffbus 3 "2000 1000 500"
  # ampbus 20 "[1.5 0]*4"
```
Interestingly it sends 20 events, one per unique effect onset times two as there's two effects to send. So it doesn't actually send the 20Hz on top as I'd expected.. Hmm!!

The /dirt/play looks like this:
![image](https://user-images.githubusercontent.com/27940/105611298-54005c80-5dac-11eb-90ab-6624fef81ac4.png)

Then the first /c_set looks like:
![image](https://user-images.githubusercontent.com/27940/105611336-7e521a00-5dac-11eb-8e79-487b5a383ab1.png)

(I haven't implemented sending multiple parameter sets in one message yet)

The rest of the /c_sets look fine..

**Comment by yaxu** [comment]
A strange thing is that if I play this:
```
d1 $ sound "sax"
  # resonance 0.2
  # cutoffbus 3 "2000 1000 500"
  # pan 1
```
Then change the pan to 0, I hear the sound panning from one side to the other, over a couple of seconds. So there's a weird lag on all the effects if I use the bus controls on one of them.

**Comment by telephon** [comment]
I would have said that `amp` isn't supported yet (because we have the formula `~amp = pow(~gain.value.min(2) + ~overgain.value, 4) * ~amp.value` in `DirtEvent`, which, in the case of `~amp = "c10"` will ignore `gain` and `overgain`.

But what you describe doesn't quite fit into this image.

**Comment by telephon** [comment]
For debugging, you could use a synth and a special parameter:

```supercollider
SynthDef(\debug, { |testvalue|
testvalue.poll(Impulse.kr(3));
}).add
```

```Haskell
d1 $ sound "debug" # testvaluebus "2000 1000 500"
```

**Comment by yaxu** [comment]
I'm afraid I have been rather stupid. 'sax' is a long sound, and I forgot to add a 'cut' or 'legato' parameter. The amp cutting the sound for everything in sync was confusing me, so I didn't realise that different instances of the sound were overlapping.

**Comment by yaxu** [comment]
Playing with `ampbus` and it seems to work fine. It's just a little bit surprising that `amp 1` is louder than no amp parameter

**Comment by yaxu** [comment]
`panbus` works fine too, but as expected is in the -1 to 1 range. I could put a hack into the scheduler to fix this.

**Comment by telephon** [comment]
> It's just a little bit surprising that `amp 1` is louder than no amp parameter

https://github.com/musikinformatik/SuperDirt/pull/220

**Comment by yaxu** [comment]
Just reading through the discussion to see what was lost in this simple implementation.. Maybe not too much, people have to handle allocating bus numbers by hand but that's OK I think.

Currently if Tidal doesn't manage to handshake straight away (i.e. if superdirt is started after tidal) then it just sends the bus ids that the user specifies, without using a mapping sent from SuperDirt. I think this works as long as superdirt is the only thing running in supercollider? The bus allocations I see are always contiguous from 0 in any case.

However it would be best to have a successful handshake in the usual cases:

- [x] Superdirt starts before Tidal
- [x] Tidal restarts while superdirt is running
- [ ] Tidal starts before superdirt
- [ ] Superdirt restarts while tidal is running 

The first two already work in Tidal 1.7. The third I've got working locally, with a simple loop waiting for a handshake.

The fourth case would be solved by this:

```
> /dirt/play "s" "bd"
< /dirt/hello
> /dirt/hello/reply tidal
> /dirt/handshake
< /dirt/handshake reply <data burst>
```

i.e. a restarted superdirt receives a `/dirt` message from a port it hasn't shaken hands with, and sends a `/dirt/hello`. Then tidal knows it has to handshake. (actually the `/dirt/hello/reply` seems redundant here, the next message could just be the handshake initiation)

Looking at the source and behaviour though the `/dirt/hello` seems to be implemented to go in the other direction?

**Comment by yaxu** [comment]
@telephon is `accelerate` intended to work with busses? I'm guessing not for the same reason as `speed`
https://github.com/tidalcycles/Tidal/pull/956/

**Comment by telephon** [comment]
No, `accelerate` doesn't work with busses. It would be complicated to implement.
